op-service/dial/active_rollup_provider.go

```diff
@@ -0,0 +1,169 @@
+package dial
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type rollupDialer func(ctx context.Context, timeout time.Duration, log log.Logger, url string) (RollupClientInterface, error)
+
+// ActiveL2EndpointProvider is an interface for providing a RollupClient
+// It manages the lifecycle of the RollupClient for callers
+// It does this by failing over down the list of rollupUrls if the current one is inactive or broken
+type ActiveL2RollupProvider struct {
+	checkDuration  time.Duration
+	networkTimeout time.Duration
+	log            log.Logger
+
+	activeTimeout time.Time
+
+	rollupUrls          []string
+	rollupDialer        rollupDialer
+	currentRollupClient RollupClientInterface
+	rollupIndex         int
+	clientLock          *sync.Mutex
+}
+
+// NewActiveL2RollupProvider creates a new ActiveL2RollupProvider
+// the checkDuration is the duration between checks to see if the current rollup client is active
+// provide a checkDuration of 0 to check every time
+func NewActiveL2RollupProvider(
+	ctx context.Context,
+	rollupUrls []string,
+	checkDuration time.Duration,
+	networkTimeout time.Duration,
+	logger log.Logger,
+) (*ActiveL2RollupProvider, error) {
+	rollupDialer := func(ctx context.Context, timeout time.Duration,
+		log log.Logger, url string,
+	) (RollupClientInterface, error) {
+		return DialRollupClientWithTimeout(ctx, timeout, log, url)
+	}
+	return newActiveL2RollupProvider(ctx, rollupUrls, checkDuration, networkTimeout, logger, rollupDialer)
+}
+
+func newActiveL2RollupProvider(
+	ctx context.Context,
+	rollupUrls []string,
+	checkDuration time.Duration,
+	networkTimeout time.Duration,
+	logger log.Logger,
+	dialer rollupDialer,
+) (*ActiveL2RollupProvider, error) {
+	if len(rollupUrls) == 0 {
```

Comment:
I wondering, do we have any test that asserts (special) behavior in the edge case where we have only a single url?

Reply:
If we have only a single url, we should be using the `StaticL2RollupProvider`/`StaticL2EndpointProvider` - nonetheless, it'd be good for the `Active` one to handle this scenario if it were somehow to occur. Adding such a test.

Reply:
Added `TestRollupProvider_HandlesSingleSequencer` and `TestEndpointProvider_HandlesSingleSequencer`